src/app/bungie-api/bungie-service-helper.ts

```diff
@@ -36,226 +41,181 @@ const notifyTimeout = _.throttle(
-    // Double the wait time, starting with 1 second, until we reach 5 minutes.
-    const waitTime = Math.min(5 * 60 * 1000, Math.pow(2, numThrottled) * 500);
-    console.log(
-      'Throttled',
-      numThrottled,
-      'times, waiting',
-      waitTime,
-      'ms before calling',
-      config.url
-    );
-    await delay(waitTime);
-  }
-
-  const whichFetch = skipAuth ? fetch : ourFetch;
-  try {
-    const result = await Promise.resolve(whichFetch(buildOptions(config, skipAuth))).then(
-      handleErrors,
-      handleErrors
-    );
-    // Quickly heal from being throttled
-    numThrottled = Math.floor(numThrottled / 2);
-    return result;
-  } catch (e) {
-    switch (e.code) {
-      case PlatformErrorCodes.ThrottleLimitExceededMinutes:
-      case PlatformErrorCodes.ThrottleLimitExceededMomentarily:
-      case PlatformErrorCodes.ThrottleLimitExceededSeconds:
-      case PlatformErrorCodes.DestinyThrottledByGameServer:
-      case PlatformErrorCodes.PerApplicationThrottleExceeded:
-      case PlatformErrorCodes.PerApplicationAnonymousThrottleExceeded:
-      case PlatformErrorCodes.PerApplicationAuthenticatedThrottleExceeded:
-      case PlatformErrorCodes.PerUserThrottleExceeded:
-      case PlatformErrorCodes.SystemDisabled:
-        numThrottled++;
-        break;
-    }
-    throw e;
-  } finally {
-    if (timer) {
-      clearTimeout(timer);
-    }
-  }
-}
-
-export function buildOptions(config: HttpClientConfig, skipAuth?: boolean): Request {
-  let url = config.url;
-  if (config.params) {
-    url = `${url}?${stringify(config.params)}`;
-  }
-
-  return new Request(url, {
-    method: config.method,
-    body: config.body ? JSON.stringify(config.body) : undefined,
-    headers: config.body
-      ? {
-          'X-API-Key': API_KEY,
-          'Content-Type': 'application/json',
-        }
-      : {
-          'X-API-Key': API_KEY,
-        },
-    credentials: skipAuth ? 'omit' : 'include',
-  });
-}
+// it would be really great if they implemented the pipeline operator soon
+/** used for most Bungie API requests */
+export const authenticatedHttpClient = dimErrorHandledHttpClient(
+  responsivelyThrottleHttpClient(
+    createHttpClient(
+      createFetchWithNonStoppingTimeout(
+        rateLimitedFetch(fetchWithBungieOAuth),
+        TIMEOUT,
+        notifyTimeout
+      ),
+      API_KEY,
+      true
+    )
+  )
+);
+/** used to get manifest and global alerts*/
+export const unauthenticatedHttpClient = dimErrorHandledHttpClient(
+  responsivelyThrottleHttpClient(
+    createHttpClient(
+      createFetchWithNonStoppingTimeout(fetch, TIMEOUT, notifyTimeout),
+      API_KEY,
+      false
+    )
+  )
+);
 
 /** Generate an error with a bit more info */
-export function error(message: string, errorCode: PlatformErrorCodes): DimError {
+export function dimError(message: string, errorCode: PlatformErrorCodes): DimError {
   const error: DimError = new Error(message);
   error.code = errorCode;
   return error;
 }
```

Comment:
I sorta feel like we should break DimError out into its own thing (a real Error subclass) and do things like pack untranslated error codes and such in there. It'd make sentry a lot easier.

Reply:
this is partly but i need to merge one thing at a time so diffs are comprehensible. dimerror stretches all over the codebase